std/range/primitives.d

```diff
@@ -1664,6 +1674,154 @@ if (isInputRange!Range)
 }
 
 /**
+This is a best-effort implementation of `back` for any kind of
+range. It expects a non-empty range.
+
+If `isBidirectional!Range`, simply returns `range.back` without
+checking `upTo) (when specified).
+
+Otherwise, walks the range through its length and returns the last element seen.
+Performes $(BIGOH n) evaluations of `range.empty` and `range.popFront()`,
+where `n` is the effective length of `range`.
+
+The `upTo` parameter is useful to "cut the losses" in case
+the interest is in seeing whether the range has at least some number
+of elements. If the parameter `upTo` is specified, stops if `upTo`
+steps have been taken and returns `upTo`.
+
+Infinite ranges are compatible, provided the parameter `upTo` is
+specified, in which case the implementation simply returns `upTo`.
+
+Params:
+    range = range to walk through
+
+Returns: last element of the range
+
+See_Also: $(REF walkLength, std, range, primitives)
+*/
+auto walkBack(Range)(Range range)
+if ((isInputRange!Range || isIterable!Range) && !isInfinite!Range)
+in
+{
+
+    assert(!range.empty, "Can't walkBack an empty range");
```

Comment:
It would be good to provide another overload which accepts a fallback value returned when the input is empty rather than causing an error